cpp/src/Slice/MetadataValidator.cpp

```diff
@@ -0,0 +1,424 @@
+        // Same thing if this metadata has been applied to a parameter or data member as well.
+        if (auto op = dynamic_pointer_cast<Operation>(p))
+        {
+            isAppliedToParameterType = true;
+            if (const auto returnType = op->returnType())
+            {
+                appliedTo = returnType;
+            }
+            else
+            {
+                string msg = '\'' + directive + "' metadata cannot be applied to operations with void return type";
+                p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg);
+                isValid = false;
+            }
+        }
+        else if (auto param = dynamic_pointer_cast<ParamDecl>(p))
+        {
+            isAppliedToParameterType = true;
+            appliedTo = param->type();
+        }
+        else if (auto dm = dynamic_pointer_cast<DataMember>(p))
+        {
+            appliedTo = dm->type();
+        }
+        else
+        {
+            // Otherwise if there's nothing special going on, and we know that the metadata was applied directly to `p`.
+            appliedTo = p;
+        }
+
+        // If this metadata is only valid in the context of parameters, issue a warning if that condition wasn't met.
+        if (info.acceptedContexts == MetadataApplicationContext::ParameterTypeReferences && !isAppliedToParameterType)
+        {
+            string msg = '\'' + directive + "' metadata can only be applied to operation parameters and return types";
+            p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg);
+            isValid = false;
+        }
+    }
+
+    // Now that we've deduced exactly what Slice element the metadata should be applied to, check that it's supported.
+    const list<reference_wrapper<const type_info>>& validOn = info.validOn;
+    if (!validOn.empty() && appliedTo) // 'appliedTo' will be null if we already found a problem and should stop.
+    {
+        auto appliedToPtr = appliedTo.get();
+        auto typeComparator = [&](reference_wrapper<const type_info> t) { return t.get() == typeid(*appliedToPtr); };
+        if (std::find_if(validOn.begin(), validOn.end(), typeComparator) == validOn.end())
+        {
+            string message = misappliedMetadataMessage(metadata, appliedTo);
+            p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, message);
+            isValid = false;
+        }
+    }
+
+    // Fifth we check if this metadata is a duplicate, i.e. has the same directive already been applied in this context?
+    if (info.mustBeUnique)
+    {
+        // 'insert' only returns `true` if the value wasn't already present in the set.
+        bool wasInserted = _seenDirectives.insert(directive).second;
+        if (!wasInserted)
+        {
+            string msg = "ignoring duplicate metadata: '" + directive + "' has already been applied in this context";
```

Comment:
Is this behavior new? Does 3.7 ignore duplicates, or let them override existing entries?